### PR TITLE
[SYCL-MLIR][NFC]: Remove getReqdLocalMemory function overloads 

### PR DESCRIPTION
**Pull request by etiotto**
Simplify `LoopInternalization` pass, remove a couple of overloads for function `getReqdLocalMemory`

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 1]
```diff
@@ -242,17 +241,26 @@ unsigned getLocalMemoryRemaining(gpu::GPUModuleOp &module,
...
-      ValueOrUnsigned::get(elemSize, builder, workGroupSize.hasElemTy<Value>());
   const unsigned numDims = accTy.getDimension();
+
+  if (workGroupSize.isKnown()) {
+    unsigned memrefReqdLocalMemory = elemSize;
+    for (unsigned dim = 0; dim < numDims; ++dim)
+      memrefReqdLocalMemory *= workGroupSize.get<unsigned>(dim);
+    return memrefReqdLocalMemory;
+  }
+
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:257]: can probably be removed.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 2]
```diff
@@ -242,17 +241,26 @@ unsigned getLocalMemoryRemaining(gpu::GPUModuleOp &module,
...
-  std::variant<Value, unsigned> reqdLocalMemory =
-      ValueOrUnsigned::get(elemSize, builder, workGroupSize.hasElemTy<Value>());
   const unsigned numDims = accTy.getDimension();
+
+  if (workGroupSize.isKnown()) {
+    unsigned memrefReqdLocalMemory = elemSize;
+    for (unsigned dim = 0; dim < numDims; ++dim)
+      memrefReqdLocalMemory *= workGroupSize.get<unsigned>(dim);
+    return memrefReqdLocalMemory;
+  }
```
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:256]: yeah my other PR hasn't merged in but still thanks for pointing it out.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 1 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:257]: Yes I kind of merged/simplified the original code and forgot that `ValueOrUnsigned::[get|mul]` return an variant containing an unsigned when given unsigned arguments.  I'll remove these section of code. Thanks.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 3]
```diff
@@ -1079,35 +1045,35 @@ void LoopInternalization::transform(FunctionOpInterface func,
...
                                     const unsigned localMemoryRemaining,
                                     DataFlowSolver &solver) {
-  // Calculate the required local memory for all accesses in
-  // loopToSharedMemref to be promoted.
+  // Calculate the total amount of shared memory needed to promote the memory
+  // accesses that were identified by the analysis.
   SmallVector<gpu::GPUFuncOp> kernels;
   funcKernelInfo.getKernelCallers(func, kernels);
+
+  const unsigned numDims = getGridDimension(func);
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:1053]: Can you please add back the assertion?
`assert(numDims > 0 && numDims <= 3 && "Dimension out of range");`

Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:1053]: `getGridDimension` contains the assertion already:

```
        unsigned dim = ty.getDimension();
        assert(dim > 0 && dim <= 3 && "Dimension out of range");
        return dim;

Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:1053]: done.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp]: this code has been removed.

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:1053]: `getGridDimension` may returns 0 if it cannot find the grid dimension, and we want to assert for that case.

Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:1046]: add this check in getCandidateFunction.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**